controlplane/kubeadm/controllers/scale.go

```diff
@@ -92,8 +92,15 @@ func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
 ) (ctrl.Result, error) {
 	logger := controlPlane.Logger()
 
-	// run preflight checks ensuring the control plane is stable before proceeding with a scale up/scale down operation; if not, wait.
-	if result, err := r.preflightChecks(ctx, controlPlane); err != nil || !result.IsZero() {
+	// Pick the Machine that we should scale down.
+	machineToDelete, err := selectMachineForScaleDown(controlPlane, outdatedMachines)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to select machine for scale down")
+	}
+
+	// Run preflight checks ensuring the control plane is stable before proceeding with a scale up/scale down operation; if not, wait.
+	// Given that we're scaling down, we can exclude the machineToDelete from the preflight checks.
+	if result, err := r.preflightChecks(ctx, controlPlane, machineToDelete); err != nil || !result.IsZero() {
```

Comment:
Unless I'm missing something, this is creating an assumption that machineToDelete will always return the same Machine. While I generally expect that to be the case, I'm not sure that it will always be the case, especially once https://github.com/kubernetes-sigs/cluster-api/pull/3948 is implemented. There could also be potential issues that crop up from backup/restore of resources.

I know in the past there were discussions related to continued reconciliation of the kubeadm config map (https://github.com/kubernetes-sigs/cluster-api/issues/2859), if that was supported, then it would be possible to improve re-entrancy for scale down by:
- forwarding etcd leadership
- issuing the Machine deletion
- returning

On the next reconciliation:
- etcd membership is updated (will need to be updated to remove members for Machines that are in a deleting state)
- kubeadm configmap is updated for ClusterConfiguration membership

Reply:
Documented the above issue more in https://github.com/kubernetes-sigs/cluster-api/issues/3978

> Unless I'm missing something, this is creating an assumption that machineToDelete will always return the same Machine.

Yeah, we discussed it a bit and although the new delete-machine annotation _could_ change the result of the machine picked for scale down (and cause the same exact issue again), we could probably live with it for now.

> I know in the past there were discussions related to continued reconciliation of the kubeadm config map (#2859), if that was supported, then it would be possible to improve re-entrancy for scale down

While we already move etcd leadership today, the problem is that the controller crashed in between the removal and issuing the Machine deletion.

I do agree we need a better re-entrancy model, although it's definitely hard to do when we're in the self-managed scenario where etcd leadership can change quite a bit and affects leader election for a controller.

Reply:
The other solution we should implement here is that for scale up/down operations we really only need to check if 

- Is the API Server available? 
  - Is there at least one pod up? 
  - Can we connect to it through the load balancer / endpoint?
- Do we have etcd quorum?

Those are really the only two operations we should check, although I'd like to follow-up on those later, once we can create two more conditions for those and check on the specific conditions.

This PR is really a quick fix for the issue that can happen most often (note that this bug affects all of v0.3.x, not just v0.3.11) and take some time to re-evaluate which checks to perform during scale up/down.

WDYT?

Reply:
I'm ok with proceeding with this as a quick fix, mainly wanted to make sure that we are tracking longer term fixes, thanks!